src/meta-srv/src/election/mysql.rs

```diff
@@ -0,0 +1,1373 @@
+        sqlx::query(sql_factory.set_idle_session_timeout_sql())
+            .execute(&mut client)
+            .await
+            .context(MySqlExecutionSnafu)?;
+
+        let tx = listen_leader_change(leader_value.clone());
+        Ok(Arc::new(Self {
+            leader_value,
+            client: Mutex::new(client),
+            is_leader: AtomicBool::new(false),
+            leader_infancy: AtomicBool::new(false),
+            leader_watcher: tx,
+            store_key_prefix,
+            candidate_lease_ttl_secs,
+            sql_set: sql_factory.build(),
+        }))
+    }
+
+    fn election_key(&self) -> String {
+        format!("{}{}", self.store_key_prefix, ELECTION_KEY)
+    }
+
+    fn candidate_root(&self) -> String {
+        format!("{}{}", self.store_key_prefix, CANDIDATES_ROOT)
+    }
+
+    fn candidate_key(&self) -> String {
+        format!("{}{}", self.candidate_root(), self.leader_value)
+    }
+}
+
+#[async_trait::async_trait]
+impl Election for MySqlElection {
+    type Leader = LeaderValue;
+
+    fn is_leader(&self) -> bool {
+        self.is_leader.load(Ordering::Relaxed)
+    }
+
+    fn in_leader_infancy(&self) -> bool {
+        self.leader_infancy
+            .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
+            .is_ok()
+    }
+
+    /// TODO(CookiePie): Split the candidate registration and keep alive logic into separate methods, so that upper layers can call them separately.
+    async fn register_candidate(&self, node_info: &MetasrvNodeInfo) -> Result<()> {
+        let key = self.candidate_key();
+        let node_info =
+            serde_json::to_string(node_info).with_context(|_| SerializeToJsonSnafu {
+                input: format!("{node_info:?}"),
+            })?;
+        let res = self
+            .put_value_with_lease(&key, &node_info, self.candidate_lease_ttl_secs)
+            .await?;
+        // May registered before, just update the lease.
+        if !res {
+            self.delete_value(&key).await?;
+            self.put_value_with_lease(&key, &node_info, self.candidate_lease_ttl_secs)
+                .await?;
+        }
+
+        // Check if the current lease has expired and renew the lease.
+        let mut keep_alive_interval =
+            tokio::time::interval(Duration::from_secs(self.candidate_lease_ttl_secs / 2));
+        loop {
+            let _ = keep_alive_interval.tick().await;
+
+            let (_, prev_expire_time, current_time, origin) = self
+                .get_value_with_lease(&key, true)
+                .await?
+                .unwrap_or_default();
+
+            ensure!(
+                prev_expire_time > current_time,
+                UnexpectedSnafu {
+                    violated: format!(
+                        "Candidate lease expired, key: {:?}",
+                        String::from_utf8_lossy(&key.into_bytes())
+                    ),
```

Comment:
Error messages could be more descriptive:

format!(Candidate lease expired at {} (current time: {}), key: {:?}", prev_expire_time, current_time, String::from_utf8_lossy(&key.into_bytes()))